File: fornituras-api/src/main/java/com/numobiz/solutions/fornituras/modules/users/service/UserService.java
package com.numobiz.solutions.fornituras.modules.users.service;

import com.numobiz.solutions.fornituras.common.audit.AuditWriter;
import com.numobiz.solutions.fornituras.common.exception.ConflictException;
import com.numobiz.solutions.fornituras.common.exception.NotFoundException;
import com.numobiz.solutions.fornituras.modules.auth.service.AuthService;
import com.numobiz.solutions.fornituras.modules.users.dto.UserRequestDTO;
import com.numobiz.solutions.fornituras.modules.users.dto.UserResponseDTO;
import com.numobiz.solutions.fornituras.modules.users.dto.UserUpdateRequest;
import com.numobiz.solutions.fornituras.modules.users.entity.Role;
import com.numobiz.solutions.fornituras.modules.users.entity.User;
import com.numobiz.solutions.fornituras.modules.users.mapper.UserMapper;
import com.numobiz.solutions.fornituras.modules.users.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@Transactional(readOnly = true)
public class UserService {

	private static final Logger log = LoggerFactory.getLogger(UserService.class);

	private final UserRepository userRepository;
	private final UserMapper userMapper;
	private final PasswordEncoder passwordEncoder;
	private final AuthService authService;
	private final AuditWriter audit;

	public UserService(
			UserRepository userRepository,
			UserMapper userMapper,
			PasswordEncoder passwordEncoder,
			AuthService authService,
			AuditWriter audit) {
		this.userRepository = userRepository;
		this.userMapper = userMapper;
		this.passwordEncoder = passwordEncoder;
		this.authService = authService;
		this.audit = audit;
	}

	public UserResponseDTO findById(Long id) {
		log.debug("Finding user by id: {}", id);
		return userMapper.toResponse(getOrThrow(id));
	}

	public Page<UserResponseDTO> findAll(Pageable pageable) {
		log.debug("Listing users (paged)");
		return userRepository.findAll(pageable).map(userMapper::toResponse);
	}

	public UserResponseDTO findByEmail(String email) {
		log.debug("Finding user by email: {}", email);
		User user = userRepository.findByEmail(email)
				.orElseThrow(() -> {
					log.warn("User not found with email: {}", email);
					return new NotFoundException("User not found with email: " + email);
				});
		return userMapper.toResponse(user);
	}

	public boolean isCurrentUser(Long id) {
		var authentication = SecurityContextHolder.getContext().getAuthentication();
		if (authentication == null || !authentication.isAuthenticated()) {
			return false;
		}
		return userRepository.findByEmail(authentication.getName())
				.map(user -> user.getId().equals(id))
				.orElse(false);
	}

	@Transactional
	public UserResponseDTO create(UserRequestDTO request) {
		log.info("Creating user with email: {}", request.email());
		if (userRepository.existsByEmail(request.email().trim())) {
			log.warn("Email already registered: {}", request.email());
			throw new ConflictException("Email already registered: " + request.email());
		}

		User user = userMapper.toEntity(request);
		user.setEmail(request.email().trim());
		user.setRole(request.role() != null ? request.role() : Role.CAPTURISTA);
		user.setEnabled(false);
		// Contraseña provisional aleatoria: el usuario define la suya al activar la cuenta con el código.
		user.setPassword(passwordEncoder.encode(UUID.randomUUID().toString() + UUID.randomUUID()));

		User savedUser = userRepository.save(user);
		authService.sendActivationCode(savedUser);
		audit.record("CREATE_USER", savedUser.getId());
		log.info("Pending user created with id: {}", savedUser.getId());
		return userMapper.toResponse(savedUser);
	}

	@Transactional
	public UserResponseDTO update(Long id, UserUpdateRequest request) {
		User user = getOrThrow(id);
		user.setName(request.name().trim());
		User saved = userRepository.save(user);
		audit.record("UPDATE_USER", saved.getId());
		return userMapper.toResponse(saved);
	}

	/**
	 * Activa/desactiva un usuario. Desactivar al <b>último administrador activo</b> se bloquea para no
	 * dejar el sistema sin control de acceso (FR-007).
	 */
	@Transactional
	public UserResponseDTO setEnabled(Long id, boolean enabled) {
		User user = getOrThrow(id);
		if (wouldLeaveSystemWithoutAdmin(user, user.getRole() == Role.ADMIN, enabled)) {
			throw new ConflictException("No se puede desactivar al último administrador activo del sistema.");
		}
		user.setEnabled(enabled);
		User saved = userRepository.save(user);
		audit.record(enabled ? "ENABLE_USER" : "DISABLE_USER", saved.getId());
		return userMapper.toResponse(saved);
	}

	/**
	 * Cambia el rol de un usuario. Degradar al <b>último administrador activo</b> a un rol no admin se
	 * bloquea (FR-007).
	 */
	@Transactional
	public UserResponseDTO changeRole(Long id, Role newRole) {
		User user = getOrThrow(id);
		if (wouldLeaveSystemWithoutAdmin(user, newRole == Role.ADMIN, user.isEnabled())) {
			throw new ConflictException("No se puede quitar el rol de administrador al último admin activo.");
		}
		user.setRole(newRole);
		User saved = userRepository.save(user);
		audit.record("ROLE_CHANGE_USER", saved.getId());
		return userMapper.toResponse(saved);
	}

	/**
	 * Determina si el estado resultante ({@code willBeAdmin}/{@code willBeEnabled}) dejaría al sistema sin
	 * ningún administrador activo: solo ocurre si el usuario es hoy admin activo, deja de serlo con el
	 * cambio y era el único administrador activo que quedaba.
	 */
	private boolean wouldLeaveSystemWithoutAdmin(User user, boolean willBeAdmin, boolean willBeEnabled) {
		boolean isCurrentlyActiveAdmin = user.getRole() == Role.ADMIN && user.isEnabled();
		if (!isCurrentlyActiveAdmin) {
			return false;
		}
		boolean staysActiveAdmin = willBeAdmin && willBeEnabled;
		return !staysActiveAdmin && userRepository.countByRoleAndEnabledTrue(Role.ADMIN) <= 1;
	}

	private User getOrThrow(Long id) {
		return userRepository.findById(id)
				.orElseThrow(() -> {
					log.warn("User not found with id: {}", id);
					return new NotFoundException("User not found with id: " + id);
				});
	}
}

File: fornituras-api-dotnet/src/Fornituras.Api/Services/UserService.cs
using Fornituras.Api.Common;
using Fornituras.Api.Data;
using Fornituras.Api.Data.Entities;
using Fornituras.Api.Dto;
using Fornituras.Api.Security;
using Microsoft.EntityFrameworkCore;

namespace Fornituras.Api.Services;

public sealed class UserService(
    ApplicationDbContext db,
    AuthService authService,
    CurrentUserService currentUser,
    IAuditWriter audit,
    ILogger<UserService> logger) : IUserService
{
    public async Task<UserResponse> FindByIdAsync(long id, CancellationToken cancellationToken = default) =>
        ToResponse(await GetOrThrowAsync(id, cancellationToken));

    public async Task<PageResult<UserResponse>> FindAllAsync(
        PaginationQuery pagination,
        string? role = null,
        bool? enabled = null,
        CancellationToken cancellationToken = default)
    {
        var query = db.Users.AsNoTracking().AsQueryable();

        if (!string.IsNullOrWhiteSpace(role))
        {
            if (!Enum.TryParse<Role>(role, true, out var parsedRole))
            {
                throw new BadRequestException($"Rol inválido: {role}");
            }

            query = query.Where(u => u.Role == parsedRole);
        }

        if (enabled is not null)
        {
            query = query.Where(u => u.Enabled == enabled);
        }

        var ordered = query.OrderBy(u => u.Id);
        var total = await ordered.CountAsync(cancellationToken);
        var items = await ordered
            .Skip(pagination.Page * pagination.Size)
            .Take(pagination.Size)
            .ToListAsync(cancellationToken);
        return PageResult<UserResponse>.From(
            items.Select(ToResponse).ToList(), total, pagination.Page, pagination.Size);
    }

    public async Task<UserResponse> FindByEmailAsync(string email, CancellationToken cancellationToken = default)
    {
        var user = await db.Users.AsNoTracking()
            .FirstOrDefaultAsync(u => u.Email == email, cancellationToken)
            ?? throw new NotFoundException($"User not found with email: {email}");
        return ToResponse(user);
    }

    public async Task<bool> IsCurrentUserAsync(long id, CancellationToken cancellationToken = default)
    {
        var email = currentUser.Email;
        if (string.IsNullOrWhiteSpace(email))
        {
            return false;
        }

        var user = await db.Users.AsNoTracking()
            .FirstOrDefaultAsync(u => u.Email == email, cancellationToken);
        return user?.Id == id;
    }

    public async Task<UserResponse> CreateAsync(UserRequest request, CancellationToken cancellationToken = default)
    {
        var email = request.Email.Trim();
        logger.LogInformation("Creating user with email: {Email}", email);

        if (await db.Users.AnyAsync(u => u.Email == email, cancellationToken))
        {
            throw new ConflictException($"Email already registered: {email}");
        }

        var role = Enum.TryParse<Role>(request.Role, true, out var parsedRole)
            ? parsedRole
            : Role.CAPTURISTA;

        var now = DateTime.UtcNow;
        var user = new User
        {
            Name = request.Name.Trim(),
            Email = email,
            Role = role,
            Enabled = false,
            Password = BCrypt.Net.BCrypt.HashPassword(Guid.NewGuid().ToString("N") + Guid.NewGuid().ToString("N")),
            CreatedAt = now,
            UpdatedAt = now
        };

        db.Users.Add(user);
        await db.SaveChangesAsync(cancellationToken);
        await authService.SendActivationCodeAsync(user, cancellationToken);
        await audit.RecordAsync("CREATE_USER", user.Id, cancellationToken);
        logger.LogInformation("Pending user created with id: {UserId}", user.Id);
        return ToResponse(user);
    }

    public async Task<UserResponse> UpdateAsync(
        long id,
        UserUpdateRequest request,
        CancellationToken cancellationToken = default)
    {
        var user = await GetOrThrowAsync(id, cancellationToken);
        user.Name = request.Name.Trim();
        user.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("UPDATE_USER", user.Id, cancellationToken);
        return ToResponse(user);
    }

    public async Task<UserResponse> SetEnabledAsync(
        long id,
        bool enabled,
        CancellationToken cancellationToken = default)
    {
        var user = await GetOrThrowAsync(id, cancellationToken);
        if (WouldLeaveSystemWithoutAdmin(user, user.Role == Role.ADMIN, enabled))
        {
            throw new ConflictException("No se puede desactivar al último administrador activo del sistema.");
        }

        user.Enabled = enabled;
        user.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync(enabled ? "ENABLE_USER" : "DISABLE_USER", user.Id, cancellationToken);
        return ToResponse(user);
    }

    public async Task<UserResponse> ChangeRoleAsync(
        long id,
        string role,
        CancellationToken cancellationToken = default)
    {
        if (!Enum.TryParse<Role>(role, true, out var newRole))
        {
            throw new BadRequestException($"Rol inválido: {role}");
        }

        return await ChangeRoleInternalAsync(id, newRole, cancellationToken);
    }

    private async Task<UserResponse> ChangeRoleInternalAsync(
        long id,
        Role newRole,
        CancellationToken cancellationToken = default)
    {
        var user = await GetOrThrowAsync(id, cancellationToken);
        if (WouldLeaveSystemWithoutAdmin(user, newRole == Role.ADMIN, user.Enabled))
        {
            throw new ConflictException("No se puede quitar el rol de administrador al último admin activo.");
        }

        user.Role = newRole;
        user.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("ROLE_CHANGE_USER", user.Id, cancellationToken);
        return ToResponse(user);
    }

    private bool WouldLeaveSystemWithoutAdmin(User user, bool willBeAdmin, bool willBeEnabled)
    {
        var isCurrentlyActiveAdmin = user.Role == Role.ADMIN && user.Enabled;
        if (!isCurrentlyActiveAdmin)
        {
            return false;
        }

        var staysActiveAdmin = willBeAdmin && willBeEnabled;
        if (staysActiveAdmin)
        {
            return false;
        }

        return db.Users.Count(u => u.Role == Role.ADMIN && u.Enabled) <= 1;
    }

    private async Task<User> GetOrThrowAsync(long id, CancellationToken cancellationToken)
    {
        return await db.Users.FirstOrDefaultAsync(u => u.Id == id, cancellationToken)
            ?? throw new NotFoundException($"User not found with id: {id}");
    }

    private static UserResponse ToResponse(User user) =>
        new(user.Id, user.Name, user.Email, user.Role.ToString(), user.Enabled, user.CreatedAt, user.UpdatedAt);
}
